biotransformers/lightning_utils/data.py

```diff
@@ -151,14 +151,38 @@ def collate_fn(
         mask_indices: indices of masked tokens
     """
     random_token_indices = [alphabet.tok_to_idx(aa) for aa in alphabet.standard_toks]
-    seqs, tokens = tokenizer(
-        samples[0]
-    )  # take samples[0] because batch_sampler return list of list
+
+    # Unpack: batch_sampler output is wrapped in a list (single item)
+    # Inside is a tuple for every sequence:
+    # First entry: entire string of the sequence;
+    # Second entry: numpy float array of weights for every character/position in the sequence.
+    list_of_tuples = samples[0]
+    sequences, weight_arrays = zip(*list_of_tuples)
+
+    # Vertically stack the weight vectors for all sequences.
+    # Note: sequences are not guaranteed to be of equal length, and thus weight arrays are not necessarily equal size.
+    # When the tokenizer is run below, the tokens_matrix will have right-padding with alphabet.padding_idx.
+    # Do the same here: right-pad with 1s to make all weight arrays equal size.
+    maximum_weight_length = max(arr.shape[0] for arr in weight_arrays)
+    weight_matrix = np.array([np.pad(arr, (0, maximum_weight_length - arr.shape[0]), mode='constant', constant_values=1.) for arr in weight_arrays])
+
+    # Add weight=1 at start and/or end,
+    # based on alphabet.prepend_bos and alphabet.append_eos,
+    # which tell us whether the tokenizer adds tokens at the beginning or end of the sequence.
+    ones_vector = np.ones((weight_matrix.shape[0], 1))
+    to_hstack = (ones_vector if alphabet.prepend_bos else None, weight_matrix, ones_vector if alphabet.append_eos else None)
+    weight_matrix = np.hstack([mat for mat in to_hstack if mat is not None])
```

Comment:
see also the [pytorch-native way](https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html) of implementing, but this looks like it also works. not sure if there's a benefit to using pytorch functions over numpy.

also while i'm nitpicking engineering details...maybe also consider parameterizing the padding methodology.